crates/ra_hir_expand/src/builtin_derive.rs

```diff
@@ -153,76 +153,105 @@ fn expand_simple_derive(
     Ok(expanded)
 }
 
+fn find_builtin_crate(db: &dyn AstDatabase, id: LazyMacroId) -> tt::TokenTree {
+    // FIXME: make hygiene works for builtin derive macro
+    // such that $crate can be used here.
+
+    let m: MacroCallId = id.into();
+    let file_id = m.as_file().original_file(db);
+    let cg = db.crate_graph();
+    let crates = db.relevant_crates(file_id);
+    let mut crate_names =
+        crates.iter().filter_map(|krate| cg[*krate].display_name.clone()).map(|it| it.to_string());
```

Comment:
hmm `display_name` really isn't supposed to be used for logic. Maybe we could just check whether the current crate has a `core` dependency?

Reply:
Yes, but I don't know a way to test if it is a `core` in this phase...

Reply:
I mean, a dependency with name `core`. All crates except `core` itself should have a dependency on `core`, so you should be able to detect `core` by seeing whether it doesn't have such a dependency.

Reply:
Okay, I change the implementation to : "checking if there is any deps, if not, it is `core` crate".

Reply:
That's not quite what I meant -- dependencies have names, so you could check whether there's a dependency named `core`. I.e. `!cg[*krate].dependencies.iter().any(|d| d.name == "core")`. It's still hacky, but maybe a bit less hacky than just checking whether there's no dependencies. Using the dependency name is better than `display_name` because the dependency name is the one actually relevant to the Rust compiler.

Reply:
Oh .. I didn't know we have `name` in `Dependency`. My bad.